### PR TITLE
fix some tricky situations around .Writing

### LINKED ISSUE #2920: restarts on single-range write txns poison txn state
consider this batch, addressed to a single range:

`BeginTransaction ["k",""), Put ["k",""), EndTransaction ["","")`

and assume `EndTransaction` fails, indicating a transaction restart.

Then the retry will fail because the transaction in the retry error takes into account the write commands before. But the transaction coordinator (for which this is the first request, so it hasn't persisted state yet), this is an error on the first attempt, so it doesn't persist the state.

That means, the client ought to be retrying with a transaction record that assumes that the writes never happened, but that's not the transaction it gets back from the error.

What it will do is use that transaction and the coordinator will error out since it looks as if the client were talking to the wrong coordinator or mishandles its transaction.

Instead we need to treat this as a read-only transaction up to that point, which will likely require some awkward manual intervention around `applyRaftCommandInBatch`.

### PR DESCRIPTION
**Pull request by tbg**
fixes #2920

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/2921)

### CODE REVIEW

File: roachpb/data.go [hunk 1]
```diff
@@ -654,10 +654,9 @@ func (t *Transaction) Update(o *Transaction) {
 	t.CertainNodes = NodeList{Nodes: append(Int32Slice(nil),
 		o.CertainNodes.Nodes...)}
 	t.UpgradePriority(o.Priority)
-	if t.Writing && !o.Writing {
-		// TODO(tschottdorf): false positives; see #2300.
-		// panic("r/w status regression")
-	} else {
+	if o.Writing {
```
Comment (bdarnell) [inline roachpb/data.go]: I think this would be a little clearer as `t.Writing = t.Writing || o.Writing`

File: roachpb/data.go [hunk 2]
```diff
@@ -623,7 +623,7 @@ func (t *Transaction) Restart(userPriority, upgradePriority int32, timestamp Tim
 // others) for the transaction. If t.ID is empty, then the transaction is
 // copied from o.
 func (t *Transaction) Update(o *Transaction) {
-	if o == nil {
+	if o == nil || len(o.ID) == 0 {
```
Comment (bdarnell) [inline roachpb/data.go]: Why did this change? When would we see a non-nil Transaction without a valid ID?

Comment (tschottdorf) [inline roachpb/data.go]: this is mostly an issue in a few tests which contain dummy errors carrying a non-nullable txn. This used to trigger the (then active) assertion below before I decided we had to remove it, so this should be fine to undo.

Comment (tschottdorf) [inline roachpb/data.go]: removed.

File: roachpb/data.go (hunk 1 above)
Comment (tschottdorf) [inline roachpb/data.go]: done.

File: kv/txn_coord_sender.go [hunk 3]
```diff
@@ -709,12 +709,18 @@ func (tc *TxnCoordSender) updateState(ctx context.Context, ba roachpb.BatchReque
 		txnMeta := tc.txns[id]
 		// For successful transactional requests, keep the written intents and
 		// the updated transaction record to be sent along with the reply.
-		// The transaction metadata is created with the first writing operation
+		// The transaction metadata is created with the first writing operation.
+		// A tricky edge case is that of a transaction which "fails" on the
+		// first writing request, but actually manages to write some intents
+		// (for example, due to being multi-range). In this case, there will
+		// be an error, but the transaction will be marked as Writing.
```
Comment (bdarnell) [inline kv/txn_coord_sender.go]: I think this comment needs more detail: why does it matter that the transaction is marked as Writing? I don't feel like i understand the issues here well enough to suggest some text.

File: kv/txn_coord_sender.go [hunk 4]
```diff
@@ -727,7 +733,7 @@ func (tc *TxnCoordSender) updateState(ctx context.Context, ba roachpb.BatchReque
 				// If the transaction is already over, there's no point in
 				// launching a one-off coordinator which will shut down right
 				// away.
-				if _, isEnding := ba.GetArg(roachpb.EndTransaction); !isEnding {
+				if _, isEnding := ba.GetArg(roachpb.EndTransaction); err == nil && !isEnding {
```
Comment (bdarnell) [inline kv/txn_coord_sender.go]: Why don't we start the heartbeat loop here? If we're keeping the `txnMeta` open in anticipation of the client's return, don't we need to heartbeat to keep the transaction alive for them?

**Comment by bdarnell** [comment]
Can you add a test for this? E.g. the scenario in #2920.

File: kv/txn_coord_sender.go (hunk 3 above)
Comment (tschottdorf) [inline kv/txn_coord_sender.go]: I'm going to update the comment.

File: kv/txn_coord_sender.go (hunk 4 above)
Comment (tschottdorf) [inline kv/txn_coord_sender.go]: You're right. Not sure what I was thinking.

**Comment by tschottdorf** [comment]
comments addressed, but I'm going to add (explicit) tests (the logic is actually exercised, but not transparently).

**Comment by bdarnell** [comment]
LGTM

**Pull request merged**